win/vs-cfg.cmd

```diff
@@ -1,175 +1,247 @@
-set "%~1=%length%"
-GOTO :EOF
+:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
+::                                                                             ::
+:: This file is part of IfcOpenShell.                                          ::
+::                                                                             ::
+:: IfcOpenShell is free software: you can redistribute it and/or modify        ::
+:: it under the terms of the Lesser GNU General Public License as published by ::
+:: the Free Software Foundation, either version 3.0 of the License, or         ::
+:: (at your option) any later version.                                         ::
+::                                                                             ::
+:: IfcOpenShell is distributed in the hope that it will be useful,             ::
+:: but WITHOUT ANY WARRANTY; without even the implied warranty of              ::
+:: MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the                ::
+:: Lesser GNU General Public License for more details.                         ::
+::                                                                             ::
+:: You should have received a copy of the Lesser GNU General Public License    ::
+:: along with this program. If not, see <http://www.gnu.org/licenses/>.        ::
+::                                                                             ::
+::::::::::::::::::::::::::::::::::::::::::::::::::::::: ::::::::::::::::::::::::::
+
+:: This script initializes various Visual Studio related environment variables.
+::
+:: It expects a CMake generator as %1. If %1 is not provided, it is deduced from
+:: the VisualStudioVersion environment variable and from the location of cl.exe.
+::
+:: The generator string must be in the CMake 3.0 and newer format, i.e.
+:: "Visual Studio 12 2013" instead of "Visual Studio 12"
+::
+:: User-friendly VS generators are preferred, since they permit a more accurate
+:: platform and toolset configuration. Some examples:
+::
+::   "vs2013"             => cmake -G "Visual Studio 12 2013" -A Win32
+::   "vs2013-x86"         => cmake -G "Visual Studio 12 2013" -A Win32
+::   "vs2015-x64"         => cmake -G "Visual Studio 14 2015" -A x64
+::   "vs2017-ARM64"       => cmake -G "Visual Studio 15 2017" -A ARM64
+::   "vs2019-x86-v141_xp" => cmake -G "Visual Studio 16 2019" -A Win32 -T v141_xp
+::
+:: NOTE: The delayed environment variable expansion needs to be enabled before calling this.
+
+@echo off
+
+set GENERATOR=%1
+
+:: Supported Visual Studio versions:
+set GENERATORS[1]="Visual Studio 12 2013"
+set GENERATORS[2]="Visual Studio 14 2015"
+set GENERATORS[3]="Visual Studio 15 2017"
+set GENERATORS[4]="Visual Studio 16 2019"
+set LAST_GENERATOR_IDX=4
+
+:: Is generator shorthand used?
+set GEN_SHORTHAND=!GENERATOR:vs=!
+
+if not "!GEN_SHORTHAND!"=="" if !GEN_SHORTHAND!==!GENERATOR! goto :GeneratorShorthandCheckDone
+set VS_PLATFORM=Win32

```

Comment:
Should we rather than defaulting to `Win32`, use %VSCMD_ARG_TGT_ARCH% which should be set by the dev tools command prompt (and maybe when that is empty) default to Win32?

Reply:
Good idea, done.